sitemedia/js/controllers/iiif_controller.js

```diff
@@ -4,32 +4,194 @@ import { Controller } from "@hotwired/stimulus";
 import OpenSeadragon from "openseadragon";
 
 export default class extends Controller {
-    static targets = ["iiifContainer"];
+    static targets = ["imageContainer"];
 
-    iiifContainerTargetConnected() {
-        // inject OSD into the iiif container
-        OpenSeadragon({
-            id: this.iiifContainerTarget.id,
+    imageContainerTargetDisconnected(container) {
+        // remove OSD on target disconnect (i.e. leaving page)
+        const image = container.querySelector("img.iiif-image");
+        this.deactivateDeepZoom(container, image);
+    }
+    handleDeepZoom(evt) {
+        // Enable OSD and/or zoom based on zoom slider level
+        const container = evt.currentTarget.parentNode.parentNode;
+        const OSD = container.querySelector(".openseadragon-container");
+        const image = container.querySelector("img.iiif-image");
+        if (!OSD || OSD.style.opacity === "0") {
+            const isMobile = evt.currentTarget.id.startsWith("zoom-toggle");
+            this.activateDeepZoom(container, image, isMobile);
+        }
+    }
+    activateDeepZoom(container, image, isMobile) {
+        // hide image and add OpenSeaDragon to container
+        const height = container.getBoundingClientRect()["height"];
+        let OSD = container.querySelector(".openseadragon-container");
+        container.style.height = `${height}px`;
+        image.classList.remove("visible");
+        image.classList.add("hidden");
+        if (!OSD) {
+            this.addOpenSeaDragon(
+                container,
+                [container.dataset.iiifUrl],
+                isMobile
+            );
+            OSD = container.querySelector(".openseadragon-container");
+        }
+        // OSD styles have to be set directly on the element instead of adding a class, due to
+        // its use of inline styles
```

Comment:
thanks for the comment. this is an OSD limitation? we can't put our styles on the container?

Reply:
wow, neither option is great. let's leave it as you've implemented for now